In [1]:
# установить свежую версию
!pip install scipy==1.5.2

     |████████████████████████████████| 25.9MB 171kB/s 
ERROR: tensorflow 2.3.0 has requirement scipy==1.4.1, but you'll have scipy 1.5.2 which is incompatible.
ERROR: albumentations 0.1.12 has requirement imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.9 which is incompatible.
  Found existing installation: scipy 1.4.1
    Uninstalling scipy-1.4.1:
      Successfully uninstalled scipy-1.4.1


In [2]:
import math
import numpy as np
import pandas as pd
from scipy import stats as st
import matplotlib.pyplot as plt 

#### Регрессионный анализ

Одним из самых распространённых видом моделей, описывающих взаимосвязь одной сл. вел. от ряда других сл. вел. или не случайных величин, является регрессионная зависимость.

Регрессионная зависимость записывается в виде линейной или нелинейной ф-ии, которая описывает данную взаимосвязь в среднем или приближенно.

Например, мы хотим выявить и описать зависимость продолжительности жизни от пола, региона проживания, уровня жизни, наличия вредных привычек и т.д. Или хотим установить зависимость дохода магазина от предлагаемого ассортимента, района расположения и т.д. Либо мы хотим выяснить, через сколько вернет кредит клиент, в зависимости от его пола, возраста, текущего уровня дохода и т.д.

$Y$ - зависимая переменная / отклик.

$X_1, ..., X_k$ - независимые переменные / факторы / предикторы 

Например, если мы рассмотрим пример с магазином, то зависимая переменная будет доход магазина, а независимые, это ассортимент, район и т.д.

На практике, строгая функциональная зависимость между сл. вел-ми практически не наблюдается, т.к. всегда возможно влияние неучтённых доп. факторов. В этом случае предполагается статистическая зависимость отклика от фактора. 

С изменением значений факторов, происходит изменение и закона распределения отклика. Мы будем рассматривать регриссионную зависимость отклика и фактора.

$y = f(x_1, ..., x_k) + \varepsilon$, где

$y$ значение отклика

$x_1, ..., x_k$ значение фактора

$f()$ функциональная взаимосвязь

$\varepsilon$ ошибка наблюдений, сл. вел. которая включает в себя неучтённые доп. факторы.

Общая задача регрессионного анализа заключается в следующем. По данным наблюдениям фактора и отклика $(x_{i1}, ..., x_{ik}, y_i)$ требуется построить оценку функциональной зависимости $y$ и $x_i$, т.е.

$\hat{y} = \hat{f}(x_1, ..., x_k)$



Следую выбранной модели $y = f(x_1, ..., x_k) + \varepsilon$ статистической зависимости, опишем модель зависимости наблюдений отклика $y_i = f(x_{i1}, ..., x_{ik}) + \varepsilon_i$, где $i = \overline {1,n}$, $n$ кол-во наблюдений.

В общих случаях, при одних и тех же значениях факторов, мы будем получать различные значения отлика $y_i$, поэтому в модель наблюдения включены ошибки наблюдений $\varepsilon_i$.

$\varepsilon_i$ - сл. величины. Будем предполагать, что

$E\varepsilon_i = 0$, $D\varepsilon_i=\sigma^2$, и они не коррелированы друг с другом $K\varepsilon_i\varepsilon_j = 0 i \neq j$

Данная модель $y = f(x_1, ..., x_k) + \varepsilon$ называется **аддитивной**.

Как правило относительно регрессионной ф-и $f$, предполагают, что её вид известен с точность до вектора неизвестных параметров $\beta$.

$\beta = (\beta_0, ..., \beta_k)$

тогда модель наблюдений примит вид

$y_i = f(x_{i1}, ..., x_{ik}; \beta) + \varepsilon_i$

наша задача, по данным наблюдениям построить оценки неизвестных параметров $\hat{\beta}$

$\hat{y_i} = f(x_{i1}, ..., x_{ik}; \hat{\beta}) + \varepsilon_i$

в итоге мы сможем построить приближенную функциональную зависимость отклика и факторов.

Например, мы предположили, что отклик зависит всего от одного фактора и эта зависимость линейна, т.е. мы имеем следующую модель наблюдения

$y_i = \beta_0 + \beta_1x_i + \varepsilon_i$

в данной модели, $\beta_0, \beta_1$ это параметры модели, они нам неизвестны, $\varepsilon_i$ это ошибка наблюдения, мы её так же не наблюдаем.

Если мы предположили, что отклик связан с одним фактором параболически, то мы будем иметь модель

$y_i = \beta_0 + \beta_1x_i + \beta_2x^2_2 + \varepsilon_i$

В случае, если регрессионная модель линейна относительно незивестных параметров $\beta_0, ..., \beta_k$ в этом случае, модель называется линейной регрессионной моделью.

Если регрессионная модель зависит нелинейно от неизвестных параметров, тогда модель называется нелинейной регрессионной моделью. Например,

$y_i = \beta^{x_i} + \varepsilon_i$

Стоит отметить, что уравнением регрессии $y$ по $x$ называют условное мат. ожидание отклика $Y$ при условии, что факторы приняли заданные значения $x_1, ..., x_k$

$E_X Y = E_X (f(x_1, ..., x_k) + \varepsilon) = f(x_1, ..., x_k)$

#### Модель простой линейно регрессии

Предположим, что зависимая переменная $Y$ зависит только от одной независимой переменной $X$.

Модель простой линейно регрессии

$y_i = \beta_0 + \beta_1x_i + \varepsilon_i$, где $i = \overline 1{1, n}$, $n$ кол-во наблюдений.

Два неизвестных параметра $\beta_0, \beta_1$, данные параметры подлежат оценки.

Предполагаем относитеьлно сл. вел. $\varepsilon_i$ предполагаем

$E\varepsilon_i = 0$, $D\varepsilon_i = \sigma^2$ дисперсия неизвестна и не наблюдаема, не коррелированы между собой $K(\varepsilon_i\varepsilon_j), i \neq j$, $\varepsilon_i \sim N(0, \sigma)$



В данном случае, отклик зависит только от одного фактора $x$, и весь разброс значений обусловлен ошибками наблюдений. Наша задача, по наблюдениям фактора и отклика, построить оценку линейной регресии

$\hat{y} = \hat{\beta_0} + \hat{\beta_1}x$

для построения точечных оценок неизвестных параметров $\beta_0, \beta_1$ как правило вводят меру рассогласования наблюдаемых значений отклика $y_i$ и предсказанных значений $\beta_0 + \beta_1x_i$.

#### Метод наименьших квадратов (МНК)

Метод построения точечных оценок неизвестных параметров $\beta_0, \beta_1$  линейной регрессионной модели.

$\sum \limits_{i=1}^n (y_i - \beta_0 - \beta_1x_i)^2 \rightarrow \underset{\beta_0, \beta_1}{min}$

Обозначим

$\sum \limits_{i=1}^n (y_i - \beta_0 - \beta_1x_i)^2 = Q$

Чтобы найти оценки, нужно найти такие $\beta_0, \beta_1$, которые минимизируют ф-ю $Q$.

$\frac{\delta Q}{\delta \beta_0} = -2 \sum \limits_{i=1}^n (y_i - \beta_0 - \beta_1x_i) = 0$

$\frac{\delta Q}{\delta \beta_1} = -2 \sum \limits_{i=1}^n (y_i - \beta_0 - \beta_1x_i) \cdot x_i = 0$

Получим систему

$\begin{equation*} 
 \begin{cases}
   n \hat{\beta_0} + \hat{\beta_1} \sum \limits_{i=1}^n x_i = \sum \limits_{i=1}^ny_i\\
   \hat{\beta_0} \sum \limits_{i=1}^n x_i + \hat{\beta_1} \sum \limits_{i=1}^n x^2_i= \sum \limits_{i=1}^n y_i x_i
 \end{cases}
\end{equation*}$

$\hat{\beta_1} = \frac{ \frac{1}{n} \sum \limits_{i=1}^n y_i x_i - \overline X \overline Y }{ \frac{1}{n} \sum \limits_{i=1}^n x^2_i - (\overline X)^2 }$

$\hat{\beta_0} = \overline Y - \hat{\beta_1} \overline X$

#### Остаточная сумма квадратов

$\hat{y}(x) = \hat{\beta_0} + \hat{\beta_1} x$

Остатки

$e_i = y_i - \hat{y}(x_i)$

$RSS = \sum \limits_{i=1}^n e^2_i = \sum \limits_{i=1}^n (y_i - \hat{y}(x_i))^2$

т.к. мат. ож. ошибок предполагаются равными нулю, то

$S^2 = \frac{RSS}{n-2} \sim \sigma^2$ $S^2$ точечная оценка дисперсии ошибок наблюдений $\sigma^2$

отметим, что $S^2$ является несмещенной оценкой ошибок наблюдений

$ES^2 = \sigma^2$

т.к. мы предполагаем, что ошибки наблюдений $\varepsilon_i \sim N(0, \sigma)$, то статистика

$Z_S = \frac{S^2 (n-2)}{\sigma^2} \sim \chi^2_{n-2}$

Статистику $RSS$ называют остаточной суммой квадратов.

#### Свойства оценок МНК $\hat{\beta_0}, \hat{\beta_1}$

1. Несмещенные. $E\hat{\beta_0} = \beta_0$, $E\hat{\beta_1} = \beta_1$

2. Дисперсии оценок равны

$D\hat{\beta_0} = \frac{\sigma^2 \sum \limits_{i=1}^n \frac{x^2_i}{n}}{n D^*_X}$

$D \hat{\beta_1} = \frac{\sigma^2}{n D^*_X}$

$D^*_X$ выборочная дисперсия икса $X$

3. т.к. у нас ошибки наблюдений $\varepsilon_i$ предполагаются нормально распределенными $\varepsilon_i \sim N(0, \sigma)$, то и оценки неизвестных параметров $\hat{\beta_0}, \hat{\beta_1}$ будут иметь так же нормальный закон распределения

$\hat{\beta_0} \sim N(\beta_0, \sqrt{D\hat{\beta_0}}), \hat{\beta_1} \sim N(\beta_1, \sqrt{D\hat{\beta_1}})$

Центрируем и нормируем оценки $\hat{\beta_0}, \hat{\beta_1}$

$\frac{\hat{\beta_i} - \beta_i}{D\hat{\beta_1}} \sim N(0, 1)$

однако дисперсия $\sigma^2$ нам неизвестна, давайте вместо неё подставим точечную оценку, т.е. $S^2$ и составим статистику

$Z_{\beta} = \frac{\sigma}{S} \frac{\hat{\beta_i} - \beta_i}{D\hat{\beta_1}} \sim T_{n-2}$

Почему у нас $n-2$ степени свободы? У нас $n$ наблюдений, но два уравнения связи, это две оценки $\hat{\beta_0}\hat{\beta_1}$, которые связывают наблюдения между собой.

#### Свойства оценок МНК $\hat{y}$

Важная задача регрессионного анализа, является построения прогноза отклика.

Прогнозом простой линейной регрессии является ф-я 

$\hat{y} = \hat{\beta_0} + \hat{\beta_1}x$

т.е. для любого $x$ мы можем построить прогноз отклика.



1. $E\hat{y} = \beta_0 + \beta_1x$
2. $D \hat{y}(x) = \frac{\sigma^2}{n} (1 + \frac{(x - \overline X)^2}{D^*_X})$

$D^*_X$ выборочная дисперсия икса $X$

3. $\hat{y} \sim (\beta_0 + \beta_1x, \sqrt{D\hat{y}(x)})$



$Z_Y = \frac{\sigma}{S} \frac{\hat{y}(x) - (\beta_0 + \beta_1x)}{\sqrt{D\hat{y}(x)}} \sim T_{n-2}$

#### Интервальные оценки

Можем построить доверительный интервал для уровня значимости $1 - \alpha$

$\beta_0$

$\hat{\beta_0} \pm t_{1 - \frac{\alpha}{2}, n-2} S \sqrt{\frac{\sum \limits_{i=1}^n \frac{x^2_i}{n}}{n D^*_X}}$



$\beta_1$

$\hat{\beta_1} \pm t_{1 - \frac{\alpha}{2}, n-2} S \sqrt{\frac{1}{n D^*_X}}$

$y(x)$

$\hat{y}(x) \pm t_{1 - \frac{\alpha}{2}, n-2} S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}$

$\sigma^2$

$(\frac{S^2(n-2)}{\chi^2_{1 - \frac{\alpha}{2}(n-2)}}, \frac{S^2(n-2)}{\chi^2_{\frac{\alpha}{2}(n-2)}})$

#### Проверка гипотез

$H_0: \beta_1 = 0$

$H_1: \beta_1 \neq 0$

Если мы сможем подтвердить гипотузу $H_0$, мы можем заключить, что фактор $x$ при этом коэфициенте не влияет на отклик $y$.

если ноль $0 \notin $ доверительному интервалу для $\hat{\beta_1}$, то $H_0$ отклоняем.

#### Задача №1

Построим линейную регрессионную модель зависимости цены автомобиля Logan и года выпуска. 

Модель: $price = \beta_0 + \beta_1 year + \varepsilon$

Найти с помощью метода наименьших квадратов оценку параметра $\beta_1$

In [5]:
df = pd.read_csv('car_regr.txt', sep='\t')
df.head(2)

price  year  mileage auto
0    250  2010     67.0   MT
1    365  2013     59.0   MT

In [6]:
from sklearn.linear_model import LinearRegression

linreg = LinearRegression()
y = df.price.values.reshape(-1, 1)
X = df.year.values.reshape(-1, 1)
linreg.fit(y=y, X=X)
linreg.coef_

array([[26.04339938]])

Найти с помощью метода наименьших квадратов оценку параметра $\beta_0$

In [7]:
linreg.intercept_

array([-52090.26932288])

Найдите остаток $e_1 = y_1 - \hat{y_1}$  соответствующий первому наблюдению

In [56]:
y1 = y[0]
x1 = df.year.values[0].reshape(-1, 1)
y1_pred = linreg.predict(x1)
e1 = y1 - y1_pred
e1

array([[-6.9634226]])

Найдите остаточную сумму квадратов RSS. $RSS = \sum \limits_{i=1}^n e^2_i = \sum \limits_{i=1}^n (y_i - \hat{y}(x_i))^2$

In [57]:
y_pred = linreg.predict(X)
RSS = np.sum((y - y_pred)**2)
RSS

132015.31925393618

Найдите $S$ (оценка дисперсии ошибок наблюдений). $S^2 = \frac{RSS}{n-2} \sim \sigma^2$ $S^2$ точечная оценка дисперсии ошибок наблюдений $\sigma^2$

In [58]:
n = df.price.count()
S_square = RSS / (n - 2)
S = np.sqrt(S_square)
S

56.06444405323646

Найдите точность 95%го доверительного интервала для параметра $\beta_1$, а именно найдите $t_{1 - \frac{\alpha}{2}, n-2} S \sqrt{\frac{1}{n D^*_{year}}}$

In [59]:
alpha = 0.05
# ddof=0 смещенная ddof=1 несмещенная
D_year = df.year.var(ddof=0)
t = st.t.ppf(1 - (alpha/2), n-2)
interval = t * S * np.sqrt(1/(n * D_year))
interval

6.393574818760041

Найдите прогноз цены  автомобиля 2012 года выпуска.

In [31]:
x_2012 = df[df.year == 2012].year.values.reshape(-1, 1)
y_pred_2012 = linreg.predict(x_2012)
y_pred_2012

array([[309.05022135],
       [309.05022135],
       [309.05022135]])

Какова вероятность того, что цена  автомобиля Logan 2012 года выпуска будет меньше 290 тыс. руб?

Ф-ю $P$ рассмотрим как наше предсказание.

Эта вероятность 

$P(P(2012) < 290)$

Будет эквивалентна такой вероятности

$P(P(2012) < 290) = P \bigg ( \frac{\hat{P}(2012) - P(2012)}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}} >  \frac{\hat{P}(2012) - 290}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}} \bigg )$, где

$S = \sqrt{S^2}$ корень из оценки дисперсии ошибки

$\overline X$ среднее выборочное значение по переменной year

$D^*_X$ выборочная дисперсия переменной year

$n$ кол-во всех наблюдений

$x$ переменная year, для которой вычисляем, в этом случае 2012

Вспомним, что мы работаем в предположении, что ошибки наблюдений имеют нормальный закон распределения $\varepsilon \sim N(a, \sigma)$. Тогда сл. вел. имеет распределение Стьюдента с $n-2$ степенями свободы 

$\frac{\hat{P}(2012) - P(2012)}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}} \sim T_{n-2}$




В правой части неравенства мы можем вычислить числовое значение, т.к. все переменные известны

$\frac{\hat{P}(2012) - 290}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}}$

In [122]:
x_2012 = df[df.year == 2012].year.values[0].reshape(-1, 1)
y_pred_2012 = linreg.predict(x_2012)
X_year_mean = df.year.mean()
# ddof=0 смещенная ddof=1 несмещенная
D_year = df.year.var(ddof=0)
# кол-во машин 2012 года
n = df.year.count()

# истинная стоимость
y = df.price.values.reshape(-1, 1)
# года
X = df.year.values.reshape(-1, 1)
# предсказанная по году стоимость
y_pred = linreg.predict(X)
# остаточная сумма квадратов
RSS = np.sum((y - y_pred)**2)

S_square = RSS / (n - 2)
S = np.sqrt(S_square)

Разобъем код по частям, чтобы не запустаться. $\frac{(x - \overline X)^2}{D^*_X}$

In [124]:
sum = (x_2012 - X_year_mean)**2 / D_year

Подкорреное выражение $\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n}$

In [125]:
root_expression = (1 + sum) / n

Всё выражение будет $\frac{\hat{P}(2012) - 290}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}} = 1.73$

In [126]:
prob = (y_pred_2012 - 290) / (S * np.sqrt(root_expression))
prob

array([[1.73746494]])

$P(P(2012) < 290) = P \bigg ( \frac{\hat{P}(2012) - P(2012)}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}} \sim T_{n-2} >  \frac{\hat{P}(2012) - 290}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}} = 1.73 \bigg )$

тогда нам достаточно вычислить значение ф-и распределения Стьюдента $1 - F_{T_{n-2}}(1.73)$



##### Важное замечание!!!

Обратим внимание, в случае, когда нужно найти вероятность, что превзойдёт какое-то значение, нужно будет вычислить ф-ю распределения стьюдента $F_{T_{n-2}}(1.73)$

$P(P(2012) > 290) = P \bigg ( \frac{\hat{P}(2012) - P(2012)}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}} \sim T_{n-2} <  \frac{\hat{P}(2012) - 290}{S \sqrt{\frac{1 + \frac{(x - \overline X)^2}{D^*_X}}{n }}} = 1.73 \bigg )$

Вычислим $1 - F_{T_{n-2}}(1.73)$

In [129]:
st.t.sf(prob, n-2)

array([[0.04481702]])

**Ответ:**

$\beta_1 = 26$

$\beta_0 = -52090$

$e_1 = -6.96$

$P(P(2012) < 290) = 0.04$

$RSS = 132015$

#### Нелинейные модели, сводящиеся к линейным

Нередно, нелинейные модели можно привести к линейной модели, путём преобразования либо фактора, либо отклика. Рассмотрим несколько примеров, начнём с преобразования факторов.


Например, мы предполагаем, что отклик связан с фактором следующим соотношением

Обратное преобразование: $Y = \beta_0 + \beta_1(1/x) + \varepsilon$

Замена $Z = 1/x$

Получаем: $Y = \beta_o + \beta_1Z + \varepsilon$

И находить оценки неизвестных параметров  $\beta_0, \beta_1$ из данной модели, далее можно обратно заменить $Z$ на $1/X$ и использовать модель для предсказания значений отклика.



Не редко, когда фактором является время, отклик связан линейно не с самим значением времни, а с логарифмом времени. В этом случае

Логарифмическое преобразование: $Y = \beta_0 + \beta_1ln(x) + \varepsilon$

Зачемна $Z = ln(x)$



Мультипликативная модель: $Y = \alpha X^{\beta} \varepsilon$

Обратим внимание, что здесь $\varepsilon$ является множителем. Мы предполагаем, что $E\varepsilon = 1$, тогда применяя натуральный  логарифм к левой части модели и правой, получаем

$ln(Y) = ln(\alpha) + \beta ln(x) + ln(\varepsilon)$, где

$ln(\alpha) = \beta_0$

$\beta = \beta_1$

$ln(Y)$ новое значение отклика

$ln(x)$ новое значение фактора

Находя оценки $\hat{\beta_0}, \hat{\beta_1}$ мы сможем из них выразить $\alpha$ и $\beta$.

В случае, если мы хотим построить дов. интерв. или проверить значимости параметров $\hat{\beta_0}, \hat{\beta_1}$ нам необходимо убедится, что ошибка наблюдений в данной модели $ln(\varepsilon) \sim N(0, \sigma)$

Обратная экспоненциальная модель: $Y = \frac{1}{1 + \alpha e^{\beta_1x + \varepsilon}}$

Применим логарифмирование к левой и правой части

$ln(1 / Y - 1) = ln(\alpha) + \beta_1x + \varepsilon$

Во всех случаях, когда мы предобразуем фактор или отклик, особенно когда мы преобразуем отклик, нам необходимо проверять, что ошибки наблюдений у нас удовлетворяют следующим требованиям

1. $E\varepsilon = 0$ они не зависят от наблюдений и имеют равные дисперсии $\sigma^2$.
2. Если мы планируем построить дов. интервалов для параметров модели или для прогноза, в этом случае нужно проверить, что ошибки наблюдений $\varepsilon \sim N(a, \sigma)$

Как проверять нормальность ошибок, мы разберем в модуле "Анализ остатков".